File: backend/app/xp.py
"""
Fórmula de XP por nível (documentada no projeto):
    XP_necessario(nivel) = 100 * nivel^1.3

`xp_for_level(n)` retorna quanto XP é necessário pra SAIR do nível n
e ir pro n+1 (curva RPG suave — cresce, mas não pune demais em
níveis altos). `level_from_xp(total)` consome XP cumulativo nível a
nível até sobrar o resto que ainda não fechou o próximo nível.
"""
import math

# ---------------------------------------------------------------------------
# PONTO DE EXTENSÃO (pós-mvp, deliberadamente adiado): decadência de XP por
# atributo, pra atributos "esquecidos" perderem XP com o tempo em vez de só
# acumular pra sempre.
#
# Quando for implementar, a atividade por atributo já é rastreável sem
# migração: junte attributes -> action_log_attributes -> action_logs.created_at
# (ver actions.py) pra achar a última ação de cada atributo, em vez de
# adicionar uma coluna `last_activity_at` redundante em `attributes`. A
# decadência entraria como uma função pura aqui (tipo `decay_xp(current_xp,
# days_inactive) -> int`), chamada sob demanda no GET do atributo (ou num job
# de manutenção no startup) — nunca recalculando o histórico de action_logs
# já gravado.
# ---------------------------------------------------------------------------


def xp_for_level(level: int) -> int:
    return round(100 * math.pow(level, 1.3))


def level_from_xp(total_xp: int) -> dict:
    level = 1
    remaining = total_xp
    while remaining >= xp_for_level(level):
        remaining -= xp_for_level(level)
        level += 1
    need = xp_for_level(level)
    pct = min(100, round((remaining / need) * 100)) if need else 0
    return {
        "level": level,
        "current_level_xp": remaining,
        "xp_for_next_level": need,
        "pct": pct,
    }

File: frontend/js/widgets/xp.js
/**
 * XP_necessario(nivel) = 100 * nivel^1.3 — xp cumulativo pra passar do
 * nível N pro N+1. Portado de kami_telas_final.html (xpForLevel/levelFromXp).
 *
 * Assunção importante (marcar se o backend discordar): current_xp
 * vindo de GET /api/nucleo/attributes é tratado aqui como XP TOTAL
 * ACUMULADO daquele atributo, não "xp dentro do nível atual". Se o
 * backend usar outra convenção, o nível calculado aqui pode divergir
 * do current_level que a API devolve — nesse caso dá pra trocar esse
 * cálculo por simplesmente exibir attr.current_level direto.
 */
export function xpForLevel(level) {
  return Math.round(100 * Math.pow(level, 1.3));
}

export function levelFromXp(totalXp) {
  let level = 1;
  let remaining = totalXp;
  while (remaining >= xpForLevel(level)) {
    remaining -= xpForLevel(level);
    level++;
  }
  const need = xpForLevel(level);
  const pct = Math.min(100, Math.round((remaining / need) * 100));
  return { level, pct, remaining, need };
}
